### PR TITLE
Rendering support for Java.

### PR DESCRIPTION
**Pull request by tarehart**

### CODE REVIEW

File: src/main/java/rlbot/manager/BotLoopRenderer.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        return botLoopMap.get(bot.getIndex());
+    }
+
+    void startPacket() {
+        builder = new FlatBufferBuilder(1000);
+    }
+
+    void finishAndSendIfDifferent() {
+        RenderPacket packet = doFinishPacket();
+        if (previousPacket == null || !previousPacket.equals(packet)) {
```
Comment (nevercast) [inline src/main/java/rlbot/manager/BotLoopRenderer.java]: packet.equals(previousPacket) ties up both of these checks. No?

File: src/main/java/rlbot/render/NamedRenderer.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+public class NamedRenderer extends Renderer {
+    public NamedRenderer(final String renderName) {
+        super(renderName.hashCode());
+    }
+
+    public void startPacket() {
+        builder = new FlatBufferBuilder(1000);
+    }
+
+    public RenderPacket finishPacket() {
```
Comment (nevercast) [inline src/main/java/rlbot/render/NamedRenderer.java]: is `finishPacket` going to be used without `finishAndSend()` ? If not, can it be private?

File: src/main/java/rlbot/render/Renderer.java [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+    private final int index;
+
+    private List<Integer> renderMessageOffsets = new LinkedList<>();
+    protected FlatBufferBuilder builder;
+
+    protected Renderer(int index) {
+        this.index = index;
+    }
+
+    protected RenderPacket doFinishPacket() {
```
Comment (nevercast) [inline src/main/java/rlbot/render/Renderer.java]: Any reason not to call this `finishPacket`? Seems inconsistent with `resetPacket`

File: src/main/java/rlbot/manager/BotLoopRenderer.java (hunk 1 above)
Comment (tarehart) [reply src/main/java/rlbot/manager/BotLoopRenderer.java]: Good point!

File: src/main/java/rlbot/render/NamedRenderer.java (hunk 2 above)
Comment (tarehart) [reply src/main/java/rlbot/render/NamedRenderer.java]: Yeah, I wanted to allow the possibility of finishing without sending. People can reuse certain renderings that way.

File: src/main/java/rlbot/render/Renderer.java (hunk 3 above)
Comment (tarehart) [reply src/main/java/rlbot/render/Renderer.java]: Yep, I want the subclass NamedRenderer to be able to have a finishPacket method with public visibility, but I don't want BotLoopRenderer to have any such method exposed to bot makers. My solution was to make this protected, and following from that, I needed to have it be named differently from the finishPacket that exists in NamedRenderer.

Comment (nevercast) [reply src/main/java/rlbot/render/Renderer.java]: Awesome! Thanks.

File: src/main/java/rlbot/render/NamedRenderer.java (hunk 2 above)
Comment (nevercast) [reply src/main/java/rlbot/render/NamedRenderer.java]: Nice. Thanks.

**Comment by nevercast** [comment]
Looks like `rlbot.flat.Color` could do with a `createColor(FlatBufferBuilder, java.awt.Color)`. Another time, another PR.

**Pull request merged**